esp8266-fastled-webserver/Fields.cpp

```diff
@@ -17,479 +17,526 @@
-  for (uint8_t i = 0; i < N; i++) {
-    Field field = fields[i];
-    if (field.name == name) {
-      return field;
+    void getPalettes(JsonVariant dst) {
+      for (uint8_t i = 0; i < paletteCount; i++) {
+        dst.add(paletteNames[i]);
+      }
     }
   }
-  return Field();
-}
-template <size_t N>
-String getFieldValue(String name, const Field (&fields)[N]) {
-  Field field = getField(name, fields);
-  if (field.getValue) {
-    return field.getValue();
-  }
-  return String();
-}
-template <size_t N>
-String setFieldValue(String name, String value, const Field (&fields)[N]) {
-  Field field = getField(name, fields);
-  if (field.setValue) {
-    return field.setValue(value);
-  }
-  return String();
-}
-template <size_t N>
-String getFieldsJson(const Field (&fields)[N]) {
-  String json = "[";
-
-  for (uint8_t i = 0; i < N; i++) {
-    Field field = fields[i];
 
-    json += "{\"name\":\"" + field.name + "\",\"label\":\"" + field.label + "\",\"type\":\"" + field.type + "\"";
+  typedef String (*FieldSetter)(String);
+  typedef String (*FieldGetter)();
+  typedef void (*FieldOptions)(JsonVariant);
+  struct Field {
+    const String name;
+    const String label;
+    Field_t type;
+    uint8_t min;
+    uint8_t max;
+    FieldGetter  getValue;
+    FieldOptions getOptions;
+    FieldSetter  setValue;
+  };
+
+  void convertToJson(const Field& field, JsonVariant dst) {
+    if (field.name.length() == 0) {
+      return;
+    }
+    if (!IsValid(field.type)) {
+      return;
+    }
 
-    if(field.getValue) {
-      if (field.type == ColorFieldType || field.type == StringFieldType || field.type == LabelFieldType) {
-        json += ",\"value\":\"" + field.getValue() + "\"";
-      }
-      else {
-        json += ",\"value\":" + field.getValue();
+    dst[F("name")]  = field.name;
+    dst[F("label")] = field.label;
+
+    dst[F("type")]  = ToString(field.type);
+    // obj[F("readonly")] = (field.setValue == nullptr);
+    if (field.getValue != nullptr) {
+      switch (field.type) {
+        case Field_t::Color: // legacy ... comma-separated string of decimals values
+        case Field_t::String:
+        case Field_t::Label:
+          dst[F("value")] = field.getValue();
+          break;
+
+        case Field_t::Boolean:
+        case Field_t::Number:
+        case Field_t::Section:
+        case Field_t::Select:
+        case Field_t::UtcOffset:
+          dst[F("value")] = field.getValue().toInt(); // TODO: fix double-conversion
+          break;
+        
+        // intentionally no default ... causes compilation warning if new enum types added w/o updating here
       }
     }
-
-    if (field.type == NumberFieldType || field.type == UtcOffsetIndexFieldType) {
-      json += ",\"min\":" + String(field.min);
-      json += ",\"max\":" + String(field.max);
+    if ((field.type == Field_t::Number) || (field.type == Field_t::UtcOffset)) {
+      dst[F("min")] = field.min;
+      dst[F("max")] = field.max;
+    }
```

Comment:
could be handled in the switch directly?
```
        case Field_t::Boolean:
        case Field_t::Section:
        case Field_t::Select:
            dst[F("value")] = field.getValue(); 
            break;
        case Field_t::UtcOffset:
        case Field_t::Number:
          dst[F("value")] = field.getValue(); 
          dst[F("min")] = field.min;
          dst[F("max")] = field.max;
          break;

```

Reply:
You are correct.  

There's some oddity with the use of flash-based strings, where they do not get optimized to single storage in flash.  (e.g., there will be two copies of "value" in flash just from the quoted code snippet).  However, I agree that transitioning to a per-type serialization, even if slightly larger resulting binary, would be worth the clarity it adds.

Can you please open a new issue to have this changed, so that each field type handles its own serialization?

Reply:
opened #219 to track this.